extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java

```diff
@@ -0,0 +1,118 @@
+			final KuduTable table, final KuduOperations operations, final boolean visibilityEnabled,
+			final Predicate<GeoWaveRow> filter, final GeoWaveRowIteratorTransformer<T> rowTransformer,
+			final boolean rowMerging) {
+		this.ranges = ranges;
+		this.adapterIds = adapterIds;
+		this.table = table;
+		this.schema = table.getSchema();
+		this.operations = operations;
+		this.visibilityEnabled = visibilityEnabled;
+		this.filter = filter;
+		this.rowTransformer = rowTransformer;
+		this.rowMerging = rowMerging;
+	}
+
+	public CloseableIterator<T> results() {
+		final List<RowResultIterator> results = new ArrayList<>();
+		Iterator<KuduRow> tmpIterator;
+		for (final SinglePartitionQueryRanges r : ranges) {
+			final byte[] partitionKey = ((r.getPartitionKey() == null) || (r.getPartitionKey().length == 0))
+					? KuduUtils.EMPTY_PARTITION_KEY
+					: r.getPartitionKey();
+			for (final ByteArrayRange range : r.getSortKeyRanges()) {
+				final byte[] start = range.getStart() != null ? range.getStart() : new byte[0];
+				final byte[] end = range.getEnd() != null ? range.getEndAsNextPrefix()
+						: new byte[] { (byte) 0xFF, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF,
+								(byte) 0xFF };
+				KuduPredicate lowerPred = KuduPredicate.newComparisonPredicate(
+						schema.getColumn(KuduField.GW_SORT_KEY.getFieldName()), ComparisonOp.GREATER_EQUAL, start);
+				KuduPredicate upperPred = KuduPredicate.newComparisonPredicate(
+						schema.getColumn(KuduField.GW_SORT_KEY.getFieldName()), ComparisonOp.LESS, end);
+				KuduPredicate partitionPred = KuduPredicate.newComparisonPredicate(
+						schema.getColumn(KuduField.GW_PARTITION_ID_KEY.getFieldName()), ComparisonOp.EQUAL,
+						partitionKey);
+				for (final short adapterId : adapterIds) {
+					KuduPredicate adapterIdPred = KuduPredicate.newComparisonPredicate(
+							schema.getColumn(KuduField.GW_ADAPTER_ID_KEY.getFieldName()), ComparisonOp.EQUAL,
+							adapterId);
+					KuduScannerBuilder scannerBuilder = operations.getScannerBuilder(table);
+					KuduScanner scanner = scannerBuilder.addPredicate(lowerPred).addPredicate(upperPred)
+							.addPredicate(partitionPred).addPredicate(adapterIdPred).build();
+					while (scanner.hasMoreRows()) {
+						try {
+							results.add(scanner.nextRows());
+						} catch (KuduException e) {
+							LOGGER.error("Error when reading rows", e);
+						}
+					}
+					try {
+						scanner.close();
+					} catch (KuduException e) {
+						LOGGER.error("Error when closing scanner", e);
+					}
+				}
+			}
+		}
+
+		if (visibilityEnabled) {
+			tmpIterator = Streams.stream(Iterators.concat(results.iterator())).map(r -> new KuduRow(r)).filter(filter)
+					.iterator();
+		} else {
+			tmpIterator = Iterators.transform(Iterators.concat(results.iterator()), r -> new KuduRow(r));
+		}
+		rowTransformer.apply(rowMerging ? new GeoWaveRowMergingIterator(tmpIterator) : (Iterator) tmpIterator);
```

Comment:
Is this cast to Iterator not necessary?

Reply:
If removed, it says 'Type mismatch: cannot convert from Iterator<KuduRow> to Iterator<GeoWaveRow>'